### PR TITLE
[Merged by Bors] - Add self-hosted runners v2

### PR DESCRIPTION
**Pull request by paulhauner**
## Issue Addressed

NA

## Proposed Changes

Carries on from #4115, with the following modifications:

1. Self-hosted runners are only enabled if `github.repository == sigp/lighthouse`.
    - This allows forks to still have Github-hosted CI.
    - This gives us a method to switch back to Github-runners if we have extended downtime on self-hosted. 
1. Does not remove any existing dependency builds for Github-hosted runners (e.g., installing the latest Rust).
1. Adds the `WATCH_HOST` environment variable which defines where we expect to find the postgres db in the `watch` tests. This should be set to `host.docker.internal` for the tests to pass on self-hosted runners.

## Additional Info

NA

### CODE REVIEW

File: .github/workflows/test-suite.yml
Comment (jimmygchen) [inline .github/workflows/test-suite.yml]: I think the `vars` context is for the config variables defined in the GitHub project, maybe we need `env` here instead? I could be wrong though.

Comment (paulhauner) [reply .github/workflows/test-suite.yml]: Yeah, thanks. There's a [some complications](https://stackoverflow.com/questions/72633058/parameterize-runs-on-value-for-github-action) about if we can use `env` variables in `runs_on` or not.

I'm going to try `SELF_HOSTED_RUNNERS` as a repository-level variable. If it works it's probably better form anyway; it allows and repo to toggle it on/off.

File: .github/workflows/test-suite.yml
Comment (paulhauner) [inline .github/workflows/test-suite.yml:133]: I've swapped from `22.04` to `latest` here. If CI fails I'll switch it back.

**Comment by paulhauner** [comment]
A list of things that are different between Github-runners and our self-hosted runners for future reference:

1. Self-hosted runners don't have `sudo`.
1. Self-hosted runners fail `rustup update` with `error: could not rename component file from '/opt/rust/toolchains/stable-x86_64-unknown-linux-gnu/etc/bash_completion.d' to '/opt/rust/tmp/bsq3a_35gvhrh01r_dir/bk': Invalid cross-device link (os error 18)` ([failed CI run](https://github.com/sigp/lighthouse/actions/runs/5549951133/jobs/10134576124?pr=4506))

**Comment by paulhauner** [comment]
We're now using `github.repository == 'sigp/lighthouse'` to enable/disable self-hosted runners. I originally wanted to use a [repository variable](https://docs.github.com/en/actions/learn-github-actions/variables#creating-configuration-variables-for-a-repository) for this purpose, however they *"... are not passed to workflows that are triggered by a pull request from a fork."*. This means that a PR from `paulhauner/lighthouse` to `sigp/lighthouse` wouldn't use self-hosted runners.

I would have liked to set an env variable at the top of the play that could have been used to de-duplicate some of the `runs_on` values. But alas, using env variables in `runs_on` is not supported (relevant issue https://github.com/actions/runner/issues/480).

**Comment by paulhauner** [comment]
For the record, this does not work:

```yaml
runs-on: ${{ github.repository == 'sigp/lighthouse' && '[self-hosted, linux, large]' || 'ubuntu-latest' }}
```

We must instead wrap the list in `fromJson`:

```yaml
runs-on: ${{ github.repository == 'sigp/lighthouse' && fromJson('["self-hosted", "linux", "large"]') || 'ubuntu-latest' }}
```

**Comment by paulhauner** [comment]
I'm having issues installing a dependency on a self-hosted runner. I can't use `apt-get` because I don't have permission but I also can't use `sudo`. I think it would be ideal if we could add dependencies using the Github Actions script, I'll block this PR until we come a decision on that.

**Comment by antondlr** [comment]
if these additions differ from the GitHub-hosted runners (i.e., install 7zip), I'll add them to the runner image. 
ideally, we keep the self-hosted and GitHub-hosted runners hot-swappable.

I'll document where and how to update the images Real Soon™️, too.

File: .github/workflows/test-suite.yml
Comment (jimmygchen) [inline .github/workflows/test-suite.yml]: we could probably remove this once this is baked into the self hosted runner image

File: .github/workflows/test-suite.yml
Comment (jimmygchen) [inline .github/workflows/test-suite.yml]: same as above

File: .github/workflows/test-suite.yml
Comment (paulhauner) [reply .github/workflows/test-suite.yml:133]: `latest` seems to work, so I'll leave it as that.

Comment (paulhauner) [reply .github/workflows/test-suite.yml:133]: Oh, it's probably just working on `latest` because we're using self-hosted runners. I might just leave it on latest anyway. It was set to `22.04` in https://github.com/sigp/lighthouse/pull/3354 and I'm not exactly sure why.

Comment (jimmygchen) [reply .github/workflows/test-suite.yml:133]: We should be good-  `latest` is the same as `22.04` now
https://github.com/actions/runner-images/blob/main/README.md?plain=1#L24

looks like debug tests didn't work in `latest` (=`20.04`) at the time of #3354 and it was switched to use 22.04 to make the job pass

File: .github/workflows/test-suite.yml
Comment (paulhauner) [reply .github/workflows/test-suite.yml]: Removed!

**Comment by paulhauner** [comment]
This is ready for review! 🎉

File: .github/workflows/test-suite.yml
Comment (jimmygchen) [inline .github/workflows/test-suite.yml]: I _think_ this may not work on GitHub hosted runners if they're not running in docker containers, perhaps we can leave it empty if `env.SELF_HOSTED_RUNNERS == false` so it defaults to `localhost`?

File: .github/workflows/test-suite.yml
Comment (jimmygchen) [inline .github/workflows/test-suite.yml]: same as above

File: .github/workflows/test-suite.yml
Comment (paulhauner) [reply .github/workflows/test-suite.yml]: Good call, I had a shot at fixing this in https://github.com/sigp/lighthouse/pull/4506/commits/ce4ed6e1e9cefa6156bfdef3abd9c1040e2d5f08.

File: .github/workflows/test-suite.yml
Comment (jimmygchen) [inline .github/workflows/test-suite.yml:93]: I think we're good with this but just something worth thinking - Is there any benefit of leaving these conditionals out? 

There are a few benefits:
- CI be always the same dep versions on self-hosted / github-hosted runners
- For Rust, the workflow will automatically update to latest version as soon as it's released, it could take an extra 15s or more to install. But then we'll know we need to update our runner image.

Downside:
- it could be a bit slower, but AFAIK `rustup update stable` just returns immediately if we're already on the same version.

Comment (paulhauner) [reply .github/workflows/test-suite.yml:93]: IIRC the self-hosted runners fail if we run that llvm install action.

Comment (paulhauner) [reply .github/workflows/test-suite.yml:93]: Same with the Rust update ones.

Comment (jimmygchen) [reply .github/workflows/test-suite.yml:93]: ahh privileges! happy to leave it as it is 👍

Comment (paulhauner) [reply .github/workflows/test-suite.yml:93]: IMO it would be ideal if we can have the images perform exactly the same. Perhaps we can make an issue to come back for it.

I suspect that we're going to eventually drift into the territory of Github-hosted runners failing, though. If we're not actually testing on the Github-hosted ones they're bound to bit-rot.

**Review by jimmygchen — APPROVED** [review]
LGTM! Woot, excited to have this merged 🔥 🚀 🎉 ⏩ ⏩ ⏩

File: .github/workflows/test-suite.yml
Comment (jimmygchen) [reply .github/workflows/test-suite.yml:93]: Yeah, that's a good point - I'm hesitant to make the workflow more complicated, but if we really want to run GitHub-hosted runners, we could still run them on PRs, and use the self-hosted ones on `staging` 😛 , but perhaps something to think about later down the line.

**Comment by paulhauner** [comment]
I'll merge this when CI passes 🎉

**Comment by paulhauner** [comment]
Yay, CI passed 🎉

bors r+

**Pull request closed without merging**
